### PR TITLE
Add custom templating to Coverage Reporting

### PR DESCRIPTION
**Pull request by karcaw**
Added code to include all resources in coverage templates
Added code to the Coverage::ResourceWrapper class to output json
Added set_template call to the Coverage class
Added table.erb and json.erb templates.

I'm pretty new to ruby, and really new to chefspec, so if there are better ways to do these things, please advise.  Also I'm not seeing much automated testing of the Coverage class, so i'm not sure where to add tests for these changes.  They work in my ad-hoc testing using the example files in the directory. What other testing should I add?

### CODE REVIEW

File: lib/chefspec/coverage.rb [hunk 1]
```diff
@@ -67,7 +69,26 @@ def add_filter(filter = nil, &block)
...
+    #
+    def set_template(file = 'human.erb')
+      [
+        ChefSpec.root.join('templates', 'coverage', file),
+        File.join(Dir.pwd,file)
+      ].each do |template|
+        if File.exist? template
+          @erb = Erubis::Eruby.new(File.read(template))
+          return
+        end
```
Comment (sethvargo) [inline lib/chefspec/coverage.rb]: I think we need to rescue the exception that ERB raises when parsing fails and print a nicer error message.

File: lib/chefspec/coverage.rb [hunk 2]
```diff
@@ -67,7 +69,26 @@ def add_filter(filter = nil, &block)
...
+    #
+    # @param [string] path
+    #   The template file to use for the output of the report
+    #
+    # @return [true]
+    #
+    def set_template(file = 'human.erb')
+      [
+        ChefSpec.root.join('templates', 'coverage', file),
+        File.join(Dir.pwd,file)
```
Comment (sethvargo) [inline lib/chefspec/coverage.rb]: `File.expand_path(file, Dir.pwd)` is more appropriate here

File: lib/chefspec/coverage.rb [hunk 3]
```diff
@@ -67,7 +69,26 @@ def add_filter(filter = nil, &block)
...
+      ].each do |template|
+        if File.exist? template
+          @erb = Erubis::Eruby.new(File.read(template))
+          return
+        end
+      end
+      raise ArgumentError, "template specified cannot be found: #{file}"
```
Comment (sethvargo) [inline lib/chefspec/coverage.rb]: If you take a look at other examples, ChefSpec has a way of defining errors that we should copy here

File: lib/chefspec/coverage.rb [hunk 4]
```diff
@@ -160,6 +180,15 @@ def to_s
         @resource.to_s
       end
 
+      def to_json
+        {
+          "source_file"=>source_file,
+          "source_line"=>source_line,
+          "touched"=>touched?,
+          "resource"=>to_s
+        }.to_json
```
Comment (sethvargo) [inline lib/chefspec/coverage.rb]: Formatting here (should have spaces around hash rockets)

File: lib/chefspec/coverage.rb [hunk 5]
```diff
@@ -67,7 +69,26 @@ def add_filter(filter = nil, &block)
...
+    #   The template file to use for the output of the report
+    #
+    # @return [true]
+    #
+    def set_template(file = 'human.erb')
+      [
+        ChefSpec.root.join('templates', 'coverage', file),
+        File.join(Dir.pwd,file)
+      ].each do |template|
+        if File.exist? template
```
Comment (sethvargo) [inline lib/chefspec/coverage.rb]: Parenthesis please

File: lib/chefspec/coverage.rb [hunk 6]
```diff
@@ -28,6 +28,8 @@ def #{name}(*args, &block)
     def initialize
       @collection = {}
       @filters    = {}
+      template = ChefSpec.root.join('templates', 'coverage', 'human.erb')
```
Comment (sethvargo) [inline lib/chefspec/coverage.rb]: I think the instance variable here should really be the template itself - it doesn't make sense to construct an ERB object that might be overwritten later by the `set_template` function.

File: lib/chefspec/coverage.rb [hunk 7]
```diff
@@ -67,7 +69,26 @@ def add_filter(filter = nil, &block)
...
+    #
+    # @return [true]
+    #
+    def set_template(file = 'human.erb')
+      [
+        ChefSpec.root.join('templates', 'coverage', file),
+        File.join(Dir.pwd,file)
+      ].each do |template|
+        if File.exist? template
+          @erb = Erubis::Eruby.new(File.read(template))
```
Comment (sethvargo) [inline lib/chefspec/coverage.rb]: As per [this comment](https://github.com/sethvargo/chefspec/pull/694/files#r58626484), I think this should just set `@template` to the right path, and then we do the ERB processing in the actual report.

File: lib/chefspec/coverage.rb (hunk 6 above)
Comment (karcaw) [inline lib/chefspec/coverage.rb]: i had considered that at one point.. fixing..

File: lib/chefspec/coverage.rb (hunk 2 above)
Comment (karcaw) [inline lib/chefspec/coverage.rb]: done

File: lib/chefspec/coverage.rb (hunk 5 above)
Comment (karcaw) [inline lib/chefspec/coverage.rb]: done.

File: lib/chefspec/coverage.rb (hunk 4 above)
Comment (karcaw) [inline lib/chefspec/coverage.rb]: fixed.

File: lib/chefspec/coverage.rb (hunk 1 above)
Comment (karcaw) [inline lib/chefspec/coverage.rb]: I will address this in the report! method, as the parsing has moved based on other comments.

**Comment by karcaw** [comment]
@sethvargo Have you looked over the new code, was I able to address all your comments?

**Pull request merged**